solidity/contracts/AcreMultiAssetVault.sol

```diff
@@ -0,0 +1,284 @@
+        address _mezoPortal,
+        address[] memory _supportedAssets
+    ) public initializer {
+        __Ownable2Step_init();
+        __Ownable_init(_owner);
+
+        if (_mezoPortal == address(0)) {
+            revert ZeroAddress();
+        }
+
+        mezoPortal = IMezoPortal(_mezoPortal);
+
+        for (uint256 i = 0; i < _supportedAssets.length; i++) {
+            _addSupportedAsset(_supportedAssets[i]);
+        }
+    }
+
+    /// @notice Adds an asset to the list of assets supported by the vault.
+    /// @dev This function can only be called by the owner.
+    /// @param asset Address of the asset to be added to the supported assets.
+    function addSupportedAsset(address asset) external onlyOwner {
+        return _addSupportedAsset(asset);
+    }
+
+    /// @notice Adds an asset to the list of assets supported by the vault.
+    /// @param asset Address of the asset to be added to the supported assets.
+    function _addSupportedAsset(address asset) internal {
+        if (asset == address(0)) {
+            revert ZeroAddress();
+        }
+        if (supportedAssets[asset]) {
+            revert AssetAlreadySupported();
+        }
+
+        emit SupportedAssetAdded(asset);
+
+        supportedAssets[asset] = true;
+    }
+
+    /// @notice Removes an asset from the list of assets supported by the vault.
+    /// @dev This function can only be called by the owner.
+    /// @param asset Address of the asset to be removed from the supported assets.
+    function removeSupportedAsset(address asset) external onlyOwner {
+        if (!supportedAssets[asset]) {
+            revert AssetNotSupported();
+        }
+
+        emit SupportedAssetRemoved(asset);
+
+        supportedAssets[asset] = false;
+    }
+
+    /// @notice Deposits an asset into the vault.
+    /// @param asset Address of the asset to be deposited.
+    /// @param amount Amount of the asset to be deposited (in the asset precision).
+    /// @return The ID of the deposit.
+    function deposit(address asset, uint256 amount) external returns (uint256) {
+        return depositFor(asset, amount, msg.sender);
+    }
+
+    /// @notice Deposits assets into the vault on behalf of a specified account,
+    ///         which will become the owner of the deposit. Only the owner of the
+    ///         deposit can withdraw the assets.
+    /// @param asset Address of the asset to be deposited.
+    /// @param amount The amount of deposit.
+    /// @param depositOwner The address of the account for which the deposit is
+    ///        being made.
+    /// @return The ID of the deposit.
+    function depositFor(
+        address asset,
+        uint256 amount,
+        address depositOwner
+    ) public returns (uint256) {
+        if (!supportedAssets[asset]) {
+            revert AssetNotSupported();
+        }
+        if (amount == 0) {
+            revert InvalidAmount(amount);
+        }
+        if (depositOwner == address(0) || depositOwner == asset) {
```

Comment:
It's still possible to use another supported asset as `depositOwner`. We may consider extending this check to:
```suggestion
        if (depositOwner == address(0) || supportedAssets[depositOwner]) {
```

Reply:
I initially thought about it, but then decided to use just the asset that is used in the call, as it's more likely user will confuse parameters of the function call and copy the same address twice, rather than they use another supported asset.
But since you raised that I included your suggestion in 690c2471a5d90efc3d50cda8ccccd43c07077ded.